File: app/src/main/java/ru/neverlands/abclient/utils/AppLog.java
package ru.neverlands.abclient.utils;

import java.util.Locale;

/**
 * Единая точка логирования для Android-кода:
 * - пишет в системный logcat (читается LogcatFileRecorder),
 * - дублирует события в FileLogger (файлы в files/Logs/...).
 *
 * Используется как безопасная замена прямых вызовов android.util.Log
 * в прикладных модулях, чтобы не дублировать ручной код двойного логирования.
 */
public final class AppLog {
    private static final String DEFAULT_CHAIN = "app_log";

    private AppLog() {
    }

    public static int d(String tag, String message) {
        int result = android.util.Log.d(tag, message);
        FileLogger.trace(resolveChain(tag), normalize(message));
        return result;
    }

    public static int d(String chain, String tag, String message) {
        int result = android.util.Log.d(tag, message);
        FileLogger.trace(resolveChain(chain), normalize(message));
        return result;
    }

    public static int i(String tag, String message) {
        int result = android.util.Log.i(tag, message);
        FileLogger.trace(resolveChain(tag), normalize(message));
        return result;
    }

    public static int i(String chain, String tag, String message) {
        int result = android.util.Log.i(tag, message);
        FileLogger.trace(resolveChain(chain), normalize(message));
        return result;
    }

    public static int w(String tag, String message) {
        int result = android.util.Log.w(tag, message);
        FileLogger.warn(resolveChain(tag), normalize(message));
        return result;
    }

    public static int w(String chain, String tag, String message) {
        int result = android.util.Log.w(tag, message);
        FileLogger.warn(resolveChain(chain), normalize(message));
        return result;
    }

    public static int w(String tag, String message, Throwable error) {
        int result = android.util.Log.w(tag, message, error);
        FileLogger.warn(resolveChain(tag), normalize(message));
        if (error != null) {
            FileLogger.error(resolveChain(tag), "warning throwable", error);
        }
        return result;
    }

    public static int w(String chain, String tag, String message, Throwable error) {
        int result = android.util.Log.w(tag, message, error);
        FileLogger.warn(resolveChain(chain), normalize(message));
        if (error != null) {
            FileLogger.error(resolveChain(chain), "warning throwable", error);
        }
        return result;
    }

    public static int e(String tag, String message) {
        int result = android.util.Log.e(tag, message);
        FileLogger.error(resolveChain(tag), normalize(message), null);
        return result;
    }

    public static int e(String chain, String tag, String message) {
        int result = android.util.Log.e(tag, message);
        FileLogger.error(resolveChain(chain), normalize(message), null);
        return result;
    }

    public static int e(String tag, String message, Throwable error) {
        int result = android.util.Log.e(tag, message, error);
        FileLogger.error(resolveChain(tag), normalize(message), error);
        return result;
    }

    public static int e(String chain, String tag, String message, Throwable error) {
        int result = android.util.Log.e(tag, message, error);
        FileLogger.error(resolveChain(chain), normalize(message), error);
        return result;
    }

    public static String getStackTraceString(Throwable error) {
        return android.util.Log.getStackTraceString(error);
    }

    private static String resolveChain(String tag) {
        if (tag == null || tag.trim().isEmpty()) {
            return DEFAULT_CHAIN;
        }
        return tag.trim().toLowerCase(Locale.ROOT);
    }

    private static String normalize(String message) {
        return message == null ? "null" : message;
    }
}

File: ABClient/AppLog.cs
using System;
using System.Diagnostics;
using ABClient.MyHelpers;

namespace ABClient
{
    /// <summary>
    /// Единая точка логирования — аналог Android AppLog.java.
    /// 
    /// Пишет в:
    ///   1. System.Diagnostics.Debug (аналог Android Logcat, видно в Output VS)
    ///   2. FileLogger (файлы Logs/Critical/*.log)
    /// 
    /// Использование:
    ///   AppLog.d("TAG", "message")           — DEBUG
    ///   AppLog.i("TAG", "message")           — INFO
    ///   AppLog.w("TAG", "message")           — WARNING
    ///   AppLog.w("TAG", "message", exception) — WARNING + exception
    ///   AppLog.e("TAG", "message")           — ERROR
    ///   AppLog.e("TAG", "message", exception) — ERROR + exception
    ///   
    /// Формат в файле: "2026-04-13 20:50:39.123 [TRACE] message"
    /// Chain (для имени файла): tag → lower + sanitize → "tag"
    /// </summary>
    internal static class AppLog
    {
        private const string DefaultChain = "app_log";

        internal static void d(string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] " + message);
            FileLogger.Trace(ResolveChain(tag), Normalize(message));
        }

        internal static void d(string chain, string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] " + message);
            FileLogger.Trace(ResolveChain(chain), Normalize(message));
        }

        internal static void i(string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] " + message);
            FileLogger.Trace(ResolveChain(tag), Normalize(message));
        }

        internal static void i(string chain, string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] " + message);
            FileLogger.Trace(ResolveChain(chain), Normalize(message));
        }

        internal static void w(string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] WARN: " + message);
            FileLogger.Warn(ResolveChain(tag), Normalize(message));
        }

        internal static void w(string chain, string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] WARN: " + message);
            FileLogger.Warn(ResolveChain(chain), Normalize(message));
        }

        internal static void w(string tag, string message, Exception error)
        {
            Debug.WriteLine("[" + tag + "] WARN: " + message + " " + error);
            FileLogger.Warn(ResolveChain(tag), Normalize(message));
            if (error != null)
            {
                FileLogger.Error(ResolveChain(tag), "warning throwable", error);
            }
        }

        internal static void w(string chain, string tag, string message, Exception error)
        {
            Debug.WriteLine("[" + tag + "] WARN: " + message + " " + error);
            FileLogger.Warn(ResolveChain(chain), Normalize(message));
            if (error != null)
            {
                FileLogger.Error(ResolveChain(chain), "warning throwable", error);
            }
        }

        internal static void e(string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] ERROR: " + message);
            FileLogger.Error(ResolveChain(tag), Normalize(message), null);
        }

        internal static void e(string chain, string tag, string message)
        {
            Debug.WriteLine("[" + tag + "] ERROR: " + message);
            FileLogger.Error(ResolveChain(chain), Normalize(message), null);
        }

        internal static void e(string tag, string message, Exception error)
        {
            Debug.WriteLine("[" + tag + "] ERROR: " + message + " " + error);
            FileLogger.Error(ResolveChain(tag), Normalize(message), error);
        }

        internal static void e(string chain, string tag, string message, Exception error)
        {
            Debug.WriteLine("[" + tag + "] ERROR: " + message + " " + error);
            FileLogger.Error(ResolveChain(chain), Normalize(message), error);
        }

        private static string ResolveChain(string tag)
        {
            if (string.IsNullOrEmpty(tag))
                return DefaultChain;
            return tag.Trim().ToLowerInvariant();
        }

        private static string Normalize(string message)
        {
            if (string.IsNullOrEmpty(message))
                return string.Empty;
            return message;
        }
    }
}